Add natural sorting

rkk1995 opened the pull request:
Natural sorting is important for documents that have file names that are numbered.

Take for example 

```
>>> items
['3000.', '4.']
>>> print(sorted(items))
['3000.', '4.']
>>> print(os_sorted(items, key = lambda i: i,reverse = False))
['4.', '3000.']
>>>
```

rkk1995 commented:
Hm, I’ll update the tests. How do you feel about the idea?

rkk1995 commented:
Yeah I’m thinking I’ll make natural a completely different setting than ascending and descending. That way they can be combined.

rkk1995 commented:
Poetry is giving me alot of problems =P

rkk1995 commented:
going to update the tests next

rkk1995 commented:
Lmk what you think

rkk1995 commented:
I fixed the lint. The test failure isn't due to these changse.

Running `poetry run pytest mkdocs_awesome_pages_plugin/tests/e2e/test_mkdocs_nav.py -k test_sections_nested_rest_dict -rx` on the tip of master has the same test failures.

lukasgeiter commented:
The test is failing due to one of the dependencies you updated. You likely forgot to run `poetry install` again after switching to master. I suggest you only add `natsort` and revert all other dependency changes.

lukasgeiter commented on mkdocs_awesome_pages_plugin/navigation.py:
```diff
@@ -1,5 +1,6 @@
 import warnings
 from pathlib import Path
+from natsort import os_sort_keygen
```
Is it a good idea to use the `os_` variant of the function? I'm a bit hesitant because the behavior can change based on which OS the docs are built.

rkk1995 replied:
Good catch. Gave some more detail in another comment further down the thread.

lukasgeiter commented on mkdocs_awesome_pages_plugin/navigation.py:
```diff
@@ -78,11 +79,13 @@ def _process_children(self, children: List[NavigationItem], collapse: bool, meta
+            return
+        key = lambda i: basename(self._get_item_path(i))
+        if sort_type == Meta.SORT_NATURAL:
+            key = os_sort_keygen(key)
```
Really nice solution using the keygen function 👏

lukasgeiter commented on mkdocs_awesome_pages_plugin/tests/e2e/test_order_and_sort.py:
```diff
@@ -0,0 +1,255 @@
+from .base import E2ETestCase
+
+
+class TestOrder(E2ETestCase):
```
Please update the name of the class to match the new filename

rkk1995 replied:
Done

lukasgeiter commented on mkdocs_awesome_pages_plugin/tests/navigation/test_order_and_sort.py:
```diff
@@ -0,0 +1,279 @@
+    def test_asc_natural(self):
+        navigation = self.createAwesomeNavigation(
+            [
+                self.page("100."),
```
What's up with the `.` suffix in the page names? Does `"100"` alone not work?

rkk1995 replied:
It runs either way. I was looking to mimick where people had pages that were numbered. I'll add a test for sanity without the `.`.

lukasgeiter commented on mkdocs_awesome_pages_plugin/tests/navigation/test_order_and_sort.py:
```diff
@@ -0,0 +1,279 @@
+    def test_section_asc_natural(self):
+        navigation = self.createAwesomeNavigation(
+            [
+                self.page("1"),
```
Did you intentionally not use 100, 20 and 3 form here on?

rkk1995 replied:
Oops. Fixed it for the `SORT_NATURAL` tests.

lukasgeiter reviewed (changes requested):
Just had a look through the code. Left a few comments, but all in all I'm pretty happy with the changes. Big thanks especially for keeping things consistent with the existing code and adding all those tests!

rkk1995 commented:
I added `natsort = ">=8.1.0"` to pyproject.toml and then ran `poetry update`.

This sets mkdocs to 1.3.1 from 1.2.3 and I think the test breaks due to that. I did `pip install mkdocs==1.2.3` and ran the test and it worked afterwards. What do you suggest doing here? If I didn't run `poetry update` the github workflow threw errors I believe.

lukasgeiter commented:
This should do the trick:
```
poetry add "natsort>=8.1.0"
```

rkk1995 commented:
Fixed stuff up. The most notable change was I mistakenly used `os_sort_keygen` (nice catch). I switched this for `natsort_keygen` which is way more generic and allows you to further customize how the sorting is done. By default `natsort_keygen`  parses numbers as integers and ignores any signs ("+","-"). 

Some resources for anyone that wants to add further natsort customization to this plugin. Shouldn't be too difficult.
https://natsort.readthedocs.io/en/5.4.1/intro.html#further-customizing-natsort

lukasgeiter commented on README.md:
```diff
@@ -402,4 +412,4 @@ If you want to contribute to the code of this project, please read the [Contribu
 [github-v1]: https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin/tree/v1
 [github-issues]: https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin/issues
-[contributing]: CONTRIBUTING.md
+[contributing]: CONTRIBUTING.md
```
Please re-add the empty line at the end here

lukasgeiter commented:
Don't worry about the failing windows-latest, pypy tests. I don't think that has anything to do with your changes. I'm guessing it's probably caused by the a new image for windows-latest.

Please do fix the formatting though. After that I'll merge your PR 🙂

lukasgeiter commented:
I've just published version 2.8.0 which includes this change. Thanks for your work! ❤️
